### DISCUSSION TITLE
Apache 2.4 + PHP 7.4 on Windows : Failed loading 'libvips-42.dll'

### DISCUSSION THREAD

**Discussion by KRousselFr** [libvips/php-vips#227] [Q&A]
Hello,

I'm facing an unexpected error while trying to use 'php-vips'.
Every attempt to use Vips from PHP under Apache fails with this exception:
` FFI\Exception: Failed loading '.../libvips-42.dll'`

Of course, I checked PHP configuration:
```
[ffi]
; FFI API restriction. Possible values:
; "preload" - enabled in CLI scripts and preloaded files (default)
; "false"   - always disabled
; "true"    - always enabled
ffi.enable=true

; List of headers files to preload, wildcard patterns allowed.
;ffi.preload=
```

I tried to investigate the problem, and it seems PHP-FFI does find but can't load VIPS DLL.

I have made this simple PHP test file to reproduce the problem:

```
<?php

$dllName = 'User32.dll';

/*	$iface = 'int MessageBoxA(unsigned int hWnd,
	                          const char* lpText,
	                          const char* lpCaption,
	                          unsigned int uType);';*/

$iface = 'int MessageBeep(unsigned int uType);';

const MB_ICONINFORMATION = 0x40;

$ffi = FFI::cdef($iface, $dllName);

/*	$res = $ffi->MessageBoxA(0,
	                         "Message de PHP-FFI.",
				 "Coucou",
				 MB_ICONINFORMATION);*/
$res = $ffi->MessageBeep(MB_ICONINFORMATION);

if ($res != 0) echo("Boing !" . PHP_EOL); else var_dump($res);

// -------------------------------

$dllName = 'C:/Users/KR/Programmation/Lib-Test.dll';

$iface = 'unsigned int get_string(char* strbuf_version, unsigned int buf_siz);';

$ffi = FFI::cdef($iface, $dllName);

$astring = FFI::new('char[256]');
$res = $ffi->get_string($astring , 256);
echo(FFI::string($astring ) . PHP_EOL);

if ($res != 0) var_dump($res);

// -------------------------------

$dllName = 'C:/VIPS/vips-dev-8.15/bin/libvips-42.dll';
$iface = 'int vips_init (const char *argv0);
          const char *vips_error_buffer (void);
          int vips_version(int flag);';

$ffi = FFI::cdef($iface, $dllName);
$ffi->vips_init('toto');

$vipsVerMaj = $ffi->vips_version(0);
$vipsVerMin = $ffi->vips_version(1);
$vipsVerRev = $ffi->vips_version(2);
echo("Version VIPS : " . $vipsVerMaj . "." . $vipsVerMin . "." . $vipsVerRev . PHP_EOL);

```

And indeed, it fails on the last `FFI::cdef` line, with the Exception I mentioned earlier, while the first two DLL calls work just fine!

I'm totally lost here, any help would be greatly appreciated.

Thanks,

NOTA: I have seen #185 which seems similar to my problem, but unfortunately OP seems to have quit before solving the question...

**Comment by jcupitt** [comment]
Hi @KRousselFr, I guess you saw https://github.com/libvips/php-vips/issues/183#issuecomment-1402149080 as well? Did those steps work?

**Comment by KRousselFr** [comment]
Hello @jcupitt 

Thank you for your quick reply.

Yes, I saw #183.
I installed Wamp server version 3.3.0, which provides me with Apache and PHP.
I downloaded VIPS v. 8.15 and installed it. I added the C/VIPS/bin directory to my system path, and can call `vips.exe` from the command prompt without problem.
'php-vips' was acquired with Composer, as recommended.

And the `watermark-text.php` example does indeed work!
Actually, it made me realize that the PHP test script I copied in my first post **does** work... But only in CLI mode!

But when using the very same files from Apache: `FFI\Exception: Failed loading 'C:/VIPS/vips-dev-8.15/bin/libvips-42.dll'`.

I just can't understand why...

**Reply by jcupitt** [reply]
PHP has separate ini files for CLI mode and Apache mode, could that be it? If you search your apache folder you should be able to find another `php.ini`. Perhaps you need to enable FFI in there?

**Reply by jcupitt** [reply]
> I added the C/VIPS/bin directory to my system path

This is probably not a good idea, unfortunately. Package DLL names can easily crash into each other and cause very hard to debug issues. 

I would just set the PATH for the apache process.

**Comment by KRousselFr** [comment]
I ensured that both `php.ini` files have the `ffi.enable=true` option.

**Comment by KRousselFr** [comment]
Removing VIPS bin directory from my Path didn't change anything, unfortunately (even after restarting Apache).

**Comment by KRousselFr** [comment]
Maybe should I ask this question to the PHP maintainers? Or to libvips'?...

**Reply by jcupitt** [reply]
I would guess it's some issue with the way your apache is handling module configuration, so I'd ask on WAMP. 

I'd need to spend an hour or so duplicating your environment and I have a deadline coming up :( sorry. I'll have a go in a week or two if you've not found a solution.

Please update this issue if you find a fix in the meantime.

**Comment by KRousselFr** [comment]
Understood, thank you very much for your time!

Best regards

**Comment by KRousselFr** [comment]
Hello @jcupitt ,

I found a workaround for this issue: everything works when the native VIPS DLLs are copied in the `C:\\Windows\\System32` directory.

Though it is is not a satisfying fix (storing stuff in the very system directory is definitely not good practice), I share it as a stopgap measure in case someone else is stuck with this problem.

It may also be a clue to help identifying the root cause and building a clean fix...

I will post a new update if I find someting better.

Best regards,

**Reply by jcupitt** [reply]
Oooof that's a desperate workaround!

I guess WAMP is unsetting PATH to improve reliability. Windows makes it very easy to accidentally swap out DLLs.

They must have an official way to add DLLs that modules need. I'd check the docs, or maybe ask the devs.

**Reply by kleisauke** [reply]
Are you testing this with PHP-FPM? You may need to set the [`clear_env`](https://www.php.net/manual/en/install.fpm.configuration.php#clear-env) configuration setting to `no`, to ensure the `PATH` environment variable is retained. Or alternatively, you could also do this in `php-fpm.conf`:
```ini
env[PATH] = C:/VIPS/vips-dev-8.15/bin
```

**Reply by KRousselFr** [reply]
Hello @kleisauke 
We don't use FPM: in our setup, PHP is directly loaded as an Apache module.
Thanks nonetheless for your suggestion.

**Comment by KRousselFr** [comment]
Hello @jcupitt 

I found a better solution: still using FFI, I call the `SetDllDirectory` function (from Windows own `kernel32.dll`) and pass it the directory where WIPS DLLs reside.
After that, everything works fine, even without adding any of VIPS DLL in any system or PATH directory.

It seems the many DLL that constitue VIPS attempt to load one another, and not having them in a directory where they are automatically searched by the system causes the problems I encouter.

Thus, I use this method in an init function of mine. This seems a good enough solution for me.
